### PR TITLE
feat(new-client): add number filter to trades(736)

### PR DESCRIPTION
**Pull request by slcfv3**
-align the number columns to right, as old client
-add filters to number column like notional
-fix a small issue of set filter

### CODE REVIEW

File: src/new-client/src/App/Trades/TradesGrid/NumFilter.tsx [hunk 1]
```diff
@@ -0,0 +1,230 @@
...
+import { usePopUpMenu } from "utils/usePopUpMenu"
+import styled from "styled-components/macro"
+import {
+  ColField,
+  comparatorConfigs,
+  ComparatorType,
+  onColFilterEnterNum,
+} from "../TradesState"
+
+export const MultiSelectWrapper = styled.span`
```
Comment (iprado78) [inline src/new-client/src/App/Trades/TradesGrid/NumFilter.tsx:11]: In this case, it's not a MultiSelect.  This is heading in the direction of becoming a generic styling component that we'll use for all three filters (number/date/set).  Don't worry about making it generic.  I could try that in my PR for the date filter.  But change the name to something like "FilterPopupWrapper".

File: src/new-client/src/App/Trades/TradesGrid/NumFilter.tsx [hunk 2]
```diff
@@ -0,0 +1,230 @@
...
+  padding: 8px 15px 5px 15px;
+  cursor: pointer;
+  display: inline-block;
+
+  @media (max-width: 400px) {
+    display: none;
+  }
+`
+
+export const MultiSelectMenu = styled.div`
```
Comment (iprado78) [inline src/new-client/src/App/Trades/TradesGrid/NumFilter.tsx:28]: "FilterPopup"

File: src/new-client/src/App/Trades/TradesGrid/NumFilter.tsx [hunk 3]
```diff
@@ -0,0 +1,230 @@
...
+  max-height: 8rem;
+  overflow-y: auto;
+  top: 0px;
+  right: 50%;
+  background-color: ${({ theme }) => theme.primary.base};
+  padding: 6px;
+  box-shadow: ${({ theme }) => theme.core.textColor} 0px 0px 0.3125rem 0px;
+`
+
+export const MultiSelectOption = styled.div<{
```
Comment (iprado78) [inline src/new-client/src/App/Trades/TradesGrid/NumFilter.tsx:41]: Unused

File: src/new-client/src/App/Trades/TradesGrid/NumFilter.tsx [hunk 4]
```diff
@@ -0,0 +1,230 @@
...
+  outline: none;
+  border: none;
+  font-size: 0.75rem;
+  width: 100%;
+  padding: 2px 0;
+  color: ${({ theme }) => theme.core.textColor};
+  border-bottom: 1.5px solid ${({ theme }) => theme.primary[5]};
+  caret-color: ${({ theme }) => theme.primary.base};
+  &:focus {
+    outline: none !important;
```
Comment (iprado78) [inline src/new-client/src/App/Trades/TradesGrid/NumFilter.tsx:130]: Eventually, we should reconsider this.  Outline is important for accessibility.  We just need to get good outline styles.

File: src/new-client/src/App/Trades/TradesGrid/NumFilter.tsx [hunk 5]
```diff
@@ -0,0 +1,230 @@
...
+  border-radius: 2px;
+  margin-bottom: 5px;
+
+  &:hover {
+    background-color: ${({ theme }) => theme.core.backgroundHoverColor};
+    font-weight: ${({ selected }) => (selected ? "bold" : "normal")};
+    text-decoration: underline;
+  }
+`
+export const DropdownWrapper = styled.div`
```
Comment (iprado78) [inline src/new-client/src/App/Trades/TradesGrid/NumFilter.tsx:61]: "FilterOptionMenuWrapper"

File: src/new-client/src/App/Trades/TradesGrid/NumFilter.tsx [hunk 6]
```diff
@@ -0,0 +1,230 @@
...
+  .dd-placeholder {
+    padding-right: 20px;
+  }
+
+  i {
+    font-size: 10px;
+  }
+`
+
+export const DropdownMenu = styled.div<{
```
Comment (iprado78) [inline src/new-client/src/App/Trades/TradesGrid/NumFilter.tsx:85]: "FilterOptionMenu"

File: src/new-client/src/App/Trades/TradesGrid/NumFilter.tsx [hunk 7]
```diff
@@ -0,0 +1,230 @@
...
+  top: 40px;
+  left: 0px;
+  background-color: ${({ theme }) => theme.primary.base};
+  padding: 6px;
+  box-shadow: 0 7px 26px 0 rgba(23, 24, 25, 0.5);
+  z-index: 100;
+  width: 100%;
+`
+
+export const DropdownOption = styled.div`
```
Comment (iprado78) [inline src/new-client/src/App/Trades/TradesGrid/NumFilter.tsx:100]: "FilterOption"

File: src/new-client/src/App/Trades/TradesGrid/NumFilter.tsx [hunk 8]
```diff
@@ -0,0 +1,230 @@
...
+    background-color: ${({ theme }) => theme.core.backgroundHoverColor};
+    font-weight: normal;
+    text-decoration: underline;
+  }
+
+  i {
+    padding-left: 20px;
+  }
+`
+const Input = styled.input`
```
Comment (iprado78) [inline src/new-client/src/App/Trades/TradesGrid/NumFilter.tsx:117]: "FilterValueInput"

File: src/new-client/src/App/Trades/TradesGrid/TableHeadCell.tsx [hunk 9]
```diff
@@ -34,42 +43,104 @@ const TableHeadCell = styled.th`
...
 export const TableHeadCellContainer: React.FC<{
   field: ColField
 }> = ({ field }) => {
   const [showFilter, setShowFilter] = useState(false)
-  const ref = useRef<HTMLDivElement>(null)
+  const ref = useRef<HTMLTableHeaderCellElement>(null)
   const { displayMenu, setDisplayMenu } = usePopUpMenu(ref)
   const tableSort = useTableSort()
   const { headerName, filterType } = colConfigs[field]
+  const numeric = filterType === "number" && field !== "tradeId"
```
Comment (iprado78) [inline src/new-client/src/App/Trades/TradesGrid/TableHeadCell.tsx:65]: If trade ids have natural arithmetic order (which in our case they do), you might want to filter by aspects of that order.  If we don't want to rely on that (maybe the nature of an idea will change the future), in maybe we make "filterType" optional and there are just some columns we don't allow filtering on.

File: src/new-client/src/App/Trades/TradesGrid/TableHeadCell.tsx [hunk 10]
```diff
@@ -34,42 +43,104 @@ const TableHeadCell = styled.th`
...
+  const [comparator, setComparator] = useState("Equals")
```
Comment (iprado78) [inline src/new-client/src/App/Trades/TradesGrid/TableHeadCell.tsx:66]: This is a bit of a smell.  We shouldn't have to care about the other applied filters (set/date) at this level.  We should just instantiate the right component (NumberFilter//SetFilter/DateFilter) based on the configured filter type for that field and let the component invoke the hook and get the options.  I realize you are solving the bug we had.  But I think we need another solution.  @josepot ?

File: src/new-client/src/App/Trades/TradesState/filterState.ts [hunk 11]
```diff
@@ -5,6 +5,48 @@ import { Trade, trades$ } from "services/trades"
...
+  InRange: "In range",
+}
+
+export interface NumFilterContent {
+  comparator: ComparatorType
+  value1: number | null
+  value2?: number | null
+}
+
+export type DistinctNums = {
```
Comment (iprado78) [inline src/new-client/src/App/Trades/TradesState/filterState.ts:33]: As we move to having different filter types, I think we need to make these types conditional.   Forex, on my branch:

```
export const colFilterMap = mapObject(colConfigs, (val) => val.filterType)

export type DistinctValues = {
  [K in ColField]: typeof colFilterMap[K] extends "set"
  ? Set<Trade[K]> : null
}```

File: src/new-client/src/App/Trades/TradesState/filterState.ts [hunk 12]
```diff
@@ -47,6 +89,11 @@ interface FilterEvent {
 interface ColFieldToggle extends FilterEvent {
   value: unknown
 }
+
+interface NumFieldToggle extends FilterEvent {
```
Comment (iprado78) [inline src/new-client/src/App/Trades/TradesState/filterState.ts:93]: I think is less a toggle than a "NumFilterSet".  Toggling "off" is the reset event, right?

File: src/new-client/src/App/Trades/TradesState/filterState.ts [hunk 13]
```diff
@@ -55,6 +102,49 @@ export const [filterResets$, onFilterReset] = createListener(
   (field: ColField) => ({ field } as FilterEvent),
 )
 
+export const [colFilterNum$, onColFilterEnterNum] = createListener(
+  (field: ColField, value: NumFilterContent) =>
+    ({ field, value } as NumFieldToggle),
+)
+
+export const numberFilters$ = mergeWithKey({
+  toggle: colFilterNum$,
```
Comment (iprado78) [inline src/new-client/src/App/Trades/TradesState/filterState.ts:111]: See above re: event type

**Review by iprado78 — CHANGES_REQUESTED** [review]

File: src/new-client/src/App/Trades/TradesState/filterState.ts (hunk 11 above)
Comment (slcfv3) [reply src/new-client/src/App/Trades/TradesState/filterState.ts:33]: Do you want me to combine different filters into one observable? Or you just want to combine different types together?

File: src/new-client/src/App/Trades/TradesGrid/NumFilter.tsx (hunk 4 above)
Comment (slcfv3) [reply src/new-client/src/App/Trades/TradesGrid/NumFilter.tsx:130]: Yeah, I just mimic the style of old client, but we can discuss it later.

File: src/new-client/src/App/Trades/TradesState/filterState.ts (hunk 11 above)
Comment (iprado78) [reply src/new-client/src/App/Trades/TradesState/filterState.ts:33]: I don't have it completely worked out.  But something like this:
```
export enum DateFilterType {
  equals = "Equals",
  greaterThan = "Greater Than",
  lessThan = "Less Than",
  notEqual = "Not Equal",
  inRange = "In Range",
}

export enum NumberFilterType {
  equals = "Equals",
  greaterThan = "Greater Than",
  greaterThanOrEqual = "Greater Than or Equal",
  lessThan = "Less Than",
  lessThanOrEqual = "Less Than or Equal",
  notEqual = "Not Equal",
  inRange = "In Range",
}

export type DistinctValues = {
  [K in ColField]: typeof colFilterMap[K] extends "set"
  ? Set<Trade[K]> : null
}

export type AppliedFilters = {
  [K in ColField]: typeof colFilterMap[K] extends "date"
    ? DateFilterType
    : typeof colFilterMap[K] extends "number"
    ? NumberFilterType
    : typeof colFilterMap[K] extends "set"
    ? Set<Trade[K]>
    : never
}```

File: src/new-client/src/App/Trades/TradesGrid/TableHeadCell.tsx [hunk 14]
```diff
@@ -36,46 +41,80 @@ const TableHeadCell = styled.th`
     display: inline-block;
   }
 `
+const FlexWrapper = styled.div<{ headerFirst: boolean }>`
```
Comment (iprado78) [inline src/new-client/src/App/Trades/TradesGrid/TableHeadCell.tsx:44]: Nice to see it was this smple

**Review by iprado78 — APPROVED** [review]
Looks good to me.  Check to see if @josepot wants any changes before merge.  Changes can also be potentially captured in my DateFilter PR that will work on top of this, which will hopefully be available by tomorrow.

**Review by josepot — APPROVED** [review]

**Pull request merged**